classes/user/Report.inc.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * @defgroup lib_pkp_classes_user
+ */
+
+/**
+ * @file lib/pkp/classes/user/Report.inc.php
+ *
+ * Copyright (c) 2003-2020 Simon Fraser University
+ * Copyright (c) 2003-2020 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
+ *
+ * @class Report
+ * @ingroup lib_pkp_classes_user
+ *
+ * @brief Generates a CSV report with basic user information given a list of users and an output stream.
+ */
+
+namespace PKP\User;
+
+class Report {
+	/** @var array An array of mappings. Each mapping is an array with two elements where:
+	 * - First: string containing the column header
+	 * - Second: callable receiving an \User and returning a ?string
+	 */
+	private $_mappings = [];
+
+	/** @var iterable The report data source, should yield /User objects */
+	private $_dataSource;
+
+	/**
+	 * Constructor
+	 * @param iterable $dataSource The data source, should yield /User objects
+	 */
+	public function __construct(iterable $dataSource)
+	{
+		\AppLocale::requireComponents(LOCALE_COMPONENT_PKP_SUBMISSION, LOCALE_COMPONENT_APP_EDITOR, LOCALE_COMPONENT_PKP_USER, LOCALE_COMPONENT_PKP_COMMON);
+		$this->_dataSource = $dataSource;
+		$this->addMappings($this->_getDefaultMappings());
+		$this->addMappings($this->_getUserGroupMappings());
+	}
+
+	/**
+	 * Retrieves the data mappings
+	 * @return array A list of mappings
+	 */
+	public function getMappings(): array
+	{
+		return $this->_mappings;
+	}
+
+	/**
+	 * Replaces the data mappings
+	 * @param array $mappings A list of mappings
+	 * @return $this
+	 */
+	public function setMappings(array $mappings): self
+	{
+		$this->_mappings = $mappings;
+		return $this;
+	}
+
+	/**
+	 * Appends mappings
+	 * @param array $mappings A list of mappings
+	 * @return $this
+	 */
+	public function addMappings(array $mappings): self
+	{
+		array_push($this->_mappings, ...$mappings);
+		return $this;
+	}
+
+	/**
+	 * Serializes the report to the given output
+	 * @param resource $output A ready to write stream
+	 */
+	public function serialize($output): void
+	{
+		// Adds BOM (byte order mark) to enforce the UTF-8 format
+		fwrite($output, "\xEF\xBB\xBF");
+
+		// Outputs column headings
+		fputcsv($output, array_map(
+			function(?string $heading): ?string
+			{
+				return \PKPString::html2text($heading);
```

Comment:
I'm not sure I'm reading it right, but it looks to me like by escaping the HTML here, the CSV will include double-escaped content. I think it would be better to escape where the data is presented -- `fputcsv` should take care of CSV's escaping needs, and HTML escaping should happen in the Smarty template (https://github.com/pkp/pkp-lib/pull/6240/files#diff-5bcedff5ad35212dc345cf262eb439cd4e78e509d4032bc8d6f0fd27f9978f2bR32).

Reply:
My idea was to remove any HTML/entity from the output, so the CSV can be easily imported/consumed into another system (e.g. email delivery service). I'm not sure the fields I selected may have HTML content, I just wanted to output safely.

Reply:
It's a good idea, but should be targeted to only columns that might contain HTML; if you're not sure about any, post here and I can confirm.

Reply:
I took a look and I think none of them will have HTML, but it was a safe assumption to support my unknown mappings :)
If you want to review, [here are the fields](https://github.com/pkp/pkp-lib/pull/6240/files#diff-752409eb9449c35619d76b53814be94664528e44146454fcd740af8b84f3aa7eR95).

Reply:
Yes, none of those support HTML -- thanks for checking!

Reply:
Ok! I've removed the escaping.